aptos-move/block-executor/src/view.rs

```diff
@@ -79,33 +89,98 @@ impl<'a, T: Transaction, X: Executable> ParallelState<'a, T, X> {
+                // scheduled for re-execution, which will re-awaken cvar here.
+                // A deadlock is not possible due to these condition variables:
+                // suppose all threads are waiting on read dependency, and consider
+                // one with lowest txn_idx. It observed a dependency, so some thread
+                // aborted dep_idx. If that abort returned execution task, by
+                // minimality (lower transactions aren't waiting), that thread would
+                // finish execution unblock txn_idx, contradiction. Otherwise,
+                // execution_idx in scheduler was lower at a time when at least the
+                // thread that aborted dep_idx was alive, and again, since lower txns
+                // than txn_idx are not blocked, so the execution of dep_idx will
+                // eventually finish and lead to unblocking txn_idx, contradiction.
+                let (lock, cvar) = &*dep_condition;
+                let mut dep_resolved = lock.lock();
+                while let DependencyStatus::Unresolved = *dep_resolved {
+                    dep_resolved = cvar.wait(dep_resolved).unwrap();
+                }
+                // dep resolved status is either resolved or execution halted.
+                matches!(*dep_resolved, DependencyStatus::Resolved)
+            },
+            DependencyResult::ExecutionHalted => false,
+            DependencyResult::Resolved => true,
+        }
+    }
+
     /// Captures a read from the VM execution, but not unresolved deltas, as in this case it is the
     /// callers responsibility to set the aggregator's base value and call fetch_data again.
-    fn fetch_data(&self, key: &T::Key, txn_idx: TxnIndex) -> ReadResult<T::Value> {
+    fn read_data_by_kind(&self, key: &T::Key, txn_idx: TxnIndex, kind: ReadKind) -> ReadResult {
         use MVDataError::*;
         use MVDataOutput::*;
 
+        if let Some(data) = self
+            .captured_reads
+            .borrow()
+            .get_by_kind(key, None, kind.clone())
+        {
+            return ReadResult::from_data_read(data);
+        }
+
         loop {
             match self.versioned_map.data().fetch_data(key, txn_idx) {
                 Ok(Versioned(version, v)) => {
-                    self.captured_reads
+                    let data_read = DataRead::Versioned(version, v.clone())
+                        .as_kind(kind)
+                        .expect("Conversion of kind from Versioned must succeed");
+
+                    if self
+                        .captured_reads
                         .borrow_mut()
-                        .push(ReadDescriptor::from_versioned(key.clone(), version));
-                    return ReadResult::Value(v);
+                        .capture_read(key.clone(), None, data_read.clone())
+                        .is_err()
+                    {
+                        // Inconsistency in recorded reads.
+                        break;
```

Comment:
breaking here (and in other places where we have an inconsistency) results in returning a halt? Is this what we want?

Reply:
not after waiting a dependency :D (would just lead to bad perf - basically making every dep wait an abort later)

Reply:
I confused myself. I think it was what we wanted, but returning proper errors in each case is the way to go.